### ISSUE TITLE
MavenPublish plugin does not honor build/libs JAR artifact in multi-module library projects, creating problems for consumers

### ISSUE THREAD

**Issue by mindlessolipsist** [gradle/gradle#20294]
Reproduction:

1. `git clone https://coramoor@bitbucket.org/cooramoor/numbers.git && cd numbers && ./gradlew assemble publishToMavenLocal && cd .. && git clone https://coramoor@bitbucket.org/cooramoor/iluvatar.git && cd iluvatar`
2. `./gradlew test assemble publishToMavenLocal`
3. `ls core/build/libs` 
> see the JAR is correctly named **iluvatar-core-2022.03.28.jar**

4. `ls ~/.m2/repository/software/visionary/iluvatar/core/2022.03.28/`
>  see the JAR is named **core-2022.03.28.jar**

As written in [this commit](https://bitbucket.org/cooramoor/iluvatar/commits/924d152206260ad930025daedc36437d2d6bc9b0), this creates a problematic situation where consumers that consume multiple libraries with a common module name like _core_ are likely to face something like:
```Execution failed for task ':authenticateWithPassword:distTar'.
> Entry authenticateWithPassword-2022.03.28/lib/core-2022.03.28.jar is a duplicate but no duplicate handling strategy has been set. Please refer to https://docs.gradle.org/7.3.3/dsl/org.gradle.api.tasks.Copy.html#org.gradle.api.tasks.Copy:duplicatesStrategy for details.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org
```

I found it extremely surprising that the MavenPublish plugin does not simply take the JAR emitted from the Jar Task and include it with the same name in its publication.

Trying the [Deferred Configuration suggestion in the docs didn't help](https://docs.gradle.org/current/userguide/publishing_maven.html#publishing_maven:deferred_configuration).

The [Publishing Overview](https://docs.gradle.org/current/userguide/publishing_setup.html#publishing_overview) leads me to believe it's perfectly reasonable to expect the output of the Jar task, i.e. "  publish the project’s production JAR file — the one produced by the jar task", to be published.

Please advise if this is a misconfiguration/documentation bug (it seems like publishing with multi-module builds should come up a lot in modern Java) or if there is an issue in the MavenPublish plugin itself.

**Comment by mindlessolipsist** [comment]
Environment:

- Arch Linux
- OpenJDK 17
- Gradle 7.4.1
- Build Scan: https://scans.gradle.com/s/g2b6ios5vibog

**Comment by mindlessolipsist** [comment]
Apologies, just had to update permissions in the BitBucket after retrying original instructions on a clean box. Have updated the repro and verified that it should work correctly on a clean setup with HTTPS access and Git.

**Comment by mindlessolipsist** [comment]
To those that run into this at 4 AM like I did and use it as a convenient stopping point, _a workaround_ (which may not seem obvious until a couple of days later):

1. copy the JARs in the build/ to their correct location in the Maven repo
2. manually edit the generated Gradle .module file to point to the correct JAR

This is of course horribly unscalable and will not obviate the need to fix this bug in Gradle itself, but it does help unblock local development.

**Comment by lepe** [comment]
I also have a project with several modules and I have been looking for ways to fix this issue. @mindlessolipsist thank you for reporting it.

**Comment by jbartok** [comment]
As stated in [the documentation of the Maven Publish Plugin](https://docs.gradle.org/current/userguide/publishing_maven.html#sec:identity_values_in_the_generated_pom), the `groupId` part of your [Maven Publication](https://docs.gradle.org/current/dsl/org.gradle.api.publish.maven.MavenPublication.html#org.gradle.api.publish.maven.MavenPublication) GAV defaults to your project's `group` property. Since you don't set that explicitly, that will [default to the folder name your project is in](https://docs.gradle.org/current/dsl/org.gradle.api.Project.html#org.gradle.api.Project:group). This is the reason it ends up as `core`.

You can influence this behavior either by specifying the group attribute for your publication or by specifying the project's group property.

Same with `artifactId` and `version`.

**Comment by mindlessolipsist** [comment]
> As stated in [the documentation of the Maven Publish Plugin](https://docs.gradle.org/current/userguide/publishing_maven.html#sec:identity_values_in_the_generated_pom), the `groupId` part of your [Maven Publication](https://docs.gradle.org/current/dsl/org.gradle.api.publish.maven.MavenPublication.html#org.gradle.api.publish.maven.MavenPublication) GAV defaults to your project's `group` property. Since you don't set that explicitly, that will [default to the folder name your project is in](https://docs.gradle.org/current/dsl/org.gradle.api.Project.html#org.gradle.api.Project:group). This is the reason it ends up as `core`.
> 
> You can influence this behavior either by specifying the group attribute for your publication or by specifying the project's group property.
> 
> Same with `artifactId` and `version`.

What kind of an answer is this? Can you **specifically explain** which part of this reply "fixes":

> I found it extremely surprising that the MavenPublish plugin does not simply take the JAR emitted from the Jar Task and include it with the same name in its publication.

>  The [Publishing Overview](https://docs.gradle.org/current/userguide/publishing_setup.html#publishing_overview) leads me to believe it's perfectly reasonable to expect the output of the Jar task, i.e. " publish the project’s production JAR file — the one produced by the jar task", to be published.

> Please advise if this is a misconfiguration/documentation bug (it seems like publishing with multi-module builds should come up a lot in modern Java) or if there is an issue in the MavenPublish plugin itself.

The clearly reproducible bug here is that the documentation explicitly states that the MavenPublish plugin will `take the output of the Jar Task.` A user can modify the JAR task, as shown in the example `build.gradle`:

`
tasks.named('jar') {
    archiveBaseName = "${rootProject.name}"
    archiveAppendix = "${project.name}"
}
`

I'm not sure why we're talking about `groupId` when the clearly surprising behavior is related to *artifactId*. 

Are you saying it is a __feature__ that the MavenPublish plugin defaults to [Project.getName()](https://docs.gradle.org/current/dsl/org.gradle.api.Project.html#org.gradle.api.Project:name) and **clobbers** whatever archive names were set in the Jar task? 

Is the assertion that this is better than:

1. Use any overrides supplied for [archiveFileName ](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.bundling.Jar.html) or its constituent parts
2. If no overrides are specified, then default to [Project.getName()](https://docs.gradle.org/current/dsl/org.gradle.api.Project.html#org.gradle.api.Project:name)

I would disagree, but I'm not an owner of the project. Nevertheless, at the very least, a decision should be made to either:

1. Make the behavior of the Maven Publish plugin honor archiveFileName overrides in Jar/other plugins
2. Clearly call out a surprising conflict like this in the documentation, eliminating misleading wording in the Publishing Overview like  

> "publish the project’s production JAR file — the one produced by the jar task"

and adding something like

> Warning: If you override the expected name of your published artifact in another plugin, such as the Jar plugin, that setting may not be used by the Maven Publish plugin.

**Comment by mindlessolipsist** [comment]
From a Code perspective, the problem appears to be that the [ArchiveTaskBasedMavenArtifact](https://github.com/gradle/gradle/blob/master/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/artifact/ArchiveTaskBasedMavenArtifact.java#L26) only knows about Classifier, if has no knowledge of Extension, BaseName, or the other documented parts of [archviveFileName](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.bundling.Jar.html#org.gradle.api.tasks.bundling.Jar:archiveFileName)

Since an entire inheritance hierarchy exists around `AbstractMavenArtifact`, it's probably deemed problematic or dangerous to modify it. But since it implements `MavenArtifact`, and entirely new instance can be made that provides the other fields.  

Most of the actual work seems to be done in [DefaultMavenPublication#populateFromComponents](https://github.com/gradle/gradle/blob/master/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java#L290), but it's pretty hard to follow without attaching a debugger exactly what is doing what. Given that [publishWithOriginalFileName](https://github.com/gradle/gradle/blob/master/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java#L766) exists, it seems very strange that [getArtifactFileName](https://github.com/gradle/gradle/blob/master/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java#L806) doesn't check, rather it is done by [getPublishedFile](https://github.com/gradle/gradle/blob/master/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java#L780) and [getPublishedUrl](https://github.com/gradle/gradle/blob/master/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publication/DefaultMavenPublication.java#L802).

This seems like a case where applying [TELL, DON'T ASK](https://martinfowler.com/bliki/TellDontAsk.html) and having `MavenArtifact` provide its name would reduce a lot of code complexity in that mammoth piece, and a new implementation that specifically looked for the pieces of `archiveFileName` could honor its settings properly.

**Comment by jbartok** [comment]
First, let's keep the tone constructive. Remember that my intention was to help you, even if I haven't done a good enough job at it. 

Second. Let me try again. Building that jar and then doing something with it, which is publishing in this case, are two independent activities. Yes, the publication will use your jar's CONTENT as its input, but it has strong conventions on where it will put it into the Maven repository and under what name (see [here](https://github.com/gradle/gradle/blob/36746b85d31f25cff0fdf7ae18e5bb1f5cc91c70/subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/publisher/AbstractMavenPublisher.java#L210)). Your publication's groupId, artifactId and version are the way to influence this location, that is why I have been talking about them. 

This is not something we at Gradle came up with, it's the actual Maven convention, called "Maven 2/3 layout", see [here](https://cwiki.apache.org/confluence/display/MAVEN/Remote+repository+layout).

**Comment by leadVisionary** [comment]
I have to say I ran into this the other day. It's pretty amazing that, 2 years later, the Maven Publishing documentation is [still factually wrong](https://docs.gradle.org/current/userguide/publishing_setup.html#publishing_overview) with the stated
```
Once the appropriate plugin has been applied, you can configure the publications and repositories. For this example, we want to publish the project’s production JAR file — the one produced by the jar task — to a custom Maven repository.
```
when that's just not what it does.

**Comment by leadVisionary** [comment]
It's also worth noting that the explanation of the  "Maven 2/3 layout" is a bit of a canard. 

For starters, it's a **Draft** wiki page that's never been finalized, written in 2016 and last modified in 2023 but never finished or "made official" with a link to published Maven documentation. 

More importantly, the Maven docs are pretty clear that GAV is an **_address_**  to retrieve an artifact, as seen here:


> Artifact coordinates are most often represented as groupId:artifactId:version, or GAV in short. (Please note that artifact coordinates > have more fields, but for brevity we still call the coordinates “GAV”, not “GAVCE”). The artifact coordinates uniquely identify an 
> artifact, but do not specify anything about its source. It is up to Maven to figure out (or you to tell Maven how to figure out) how and > from where to retrieve the artifact.

Like a pointer, this address doesn't seem to mandate anything about the actual file name in the repository. The code itself in [org.eclipse.aether.artifact.Artifact ](https://github.com/apache/maven-resolver/blob/master/maven-resolver-api/src/main/java/org/eclipse/aether/artifact/Artifact.java#L115) seems to imply this pretty heavily in both the deprecated `getFile` and `getPath` with the repeated admonition 

> In general, callers must not assume any relationship between an artifact's filename and its coordinates.`

So it seems pretty shaky to assert that clobbering a Gradle user's modifications to the JAR filename is solely due to "compliance with GAV"

**Comment by leadVisionary** [comment]
In 2026, the convention is so hard-coded that the only path through `mavenLocal()` resolution for two artifacts sharing an `artifactId` from different groups  is to publish  **two byte-identical jars** —
`<artifactId>-<v>.jar` to satisfy `MavenLocalMetadataValidator`'s
synthesized `metadata.artifact("jar","jar",null)` lookup, and
`<artifactId>-<v>-<group>.jar` as the Gradle Module Metadata-redirected sidecar so
`installDist` sees distinct filenames. 

Interestingly, the validator never reads the file contents, just `Files.exists()` at the conventional path; 
GMM drives the actual classpath. We worked around the doubled storage on
disk with `Files.createLink` between the two filenames:

```
~/.m2/.../api/2026.05.10/:
  47324177  -rw-r--r-- 2 ...  api-2026.05.10.jar
  47324177  -rw-r--r-- 2 ...  api-2026.05.10-com.example.foo.jar
                     ^ refcount=2, one inode, two directory entries
```

Same inode, real `.class` files in both names, `installDist` happy.
Cool. We saved the bytes on a developer machine. 

One can only wonder how many people are saying "the hell with it, storage is cheap"
and just emitting the same JAR twice. Multiplied across Maven Central,
 (~5M GAV triples × ~500 KB median × ~10% "common-name"
collision rate) this convention could be costing up to an estimated 2.5 TB of duplicated
JAR uploads, plus ~5-20 GB per dev machine that adopts the
pattern. Maybe exabytes globally if every publisher does the obvious
fix. 

A single optional field in GMM —

```json
"files": [
  { "name": "api-1.0.jar", "linkTo": "api-1.0-com.example.foo.jar", ... },
  { "name": "api-1.0-com.example.foo.jar", "size": ..., "sha512": "..." }
]
```

— would let a `linkTo`-aware resolver skip the duplicate download and
materialise the local cache entry as a hardlink. Instead, we're hand-rolling
filesystem-level workarounds in 2026 because the _artifact resolution model_
forgot that "group" is part of identity at the layer that copies files, 
literally treating _artifactIds_ as globally unique. Fantastic.

Here's a cute little repo if the issue isn't clear:
```python
#!/usr/bin/env python3
"""
gradle/gradle#20294 reproducer.

Two publishers share an artifactId; one consumer depends on both; installDist
trips on the duplicate filename. Self-bootstrapping: pulls SDKMAN + Java +
Gradle into ~/.sdkman if missing, drives the repro from a hermetic /tmp tree.
"""
from __future__ import annotations
import os, shutil, subprocess, sys
from pathlib import Path
from textwrap import dedent

WORK   = Path("/tmp/gradle-20294-repro")
GRADLE = "9.5.0"
JAVA   = "21.0.5-tem"
SDKMAN = Path.home() / ".sdkman"


def banner(s: str) -> None:
    print(f"\n\033[1;36m── {s} ──\033[0m")


def bash(cmd: str) -> None:
    subprocess.run(["bash", "-c", cmd], check=True)


def sdk(args: str) -> None:
    bash(f"source '{SDKMAN}/bin/sdkman-init.sh' && yes | sdk {args}")


def bootstrap() -> Path:
    if not SDKMAN.exists():
        banner("installing SDKMAN")
        bash("curl -fsSL 'https://get.sdkman.io?rcupdate=false' | bash")
    if not (SDKMAN / "candidates/java/current").exists():
        banner(f"installing Java {JAVA}")
        sdk(f"install java {JAVA}")
    gradle = SDKMAN / f"candidates/gradle/{GRADLE}/bin/gradle"
    if not gradle.exists():
        banner(f"installing Gradle {GRADLE}")
        sdk(f"install gradle {GRADLE}")
    return gradle


def write(path: Path, body: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(dedent(body))


def publisher(name: str) -> Path:
    root = WORK / name
    write(root / "settings.gradle.kts", 'rootProject.name = "api"\n')
    write(root / "build.gradle.kts", f"""
        plugins {{ java; `maven-publish` }}
        group = "com.example.{name}"
        version = "1.0.0"
        publishing {{
            publications {{ create<MavenPublication>("lib") {{ from(components["java"]) }} }}
            repositories {{ maven {{ url = uri("{WORK}/m2") }} }}
        }}
    """)
    write(root / "src/main/java/example/Foo.java",
          "package example; public class Foo {}\n")
    return root


def consumer() -> Path:
    root = WORK / "app"
    write(root / "settings.gradle.kts", f"""
        dependencyResolutionManagement {{
            repositories {{ maven {{ url = uri("{WORK}/m2") }} }}
        }}
        rootProject.name = "consumer"
    """)
    write(root / "build.gradle.kts", """
        plugins {{ application }}
        dependencies {{
            implementation("com.example.foo:api:1.0.0")
            implementation("com.example.bar:api:1.0.0")
        }}
        application {{ mainClass.set("example.Main") }}
    """.replace("{{", "{").replace("}}", "}"))
    write(root / "src/main/java/example/Main.java",
          "package example;\npublic class Main { public static void main(String[] a) {} }\n")
    return root


def gradle(at: Path, bin_: Path, *args: str) -> subprocess.CompletedProcess:
    env = {
        **os.environ,
        "JAVA_HOME":        str(SDKMAN / "candidates/java/current"),
        "GRADLE_USER_HOME": str(WORK / ".gradle"),
    }
    return subprocess.run(
        [str(bin_), "--no-daemon", *args],
        cwd=at, env=env, text=True, capture_output=True,
    )


def main() -> int:
    if WORK.exists():
        shutil.rmtree(WORK)
    (WORK / "m2").mkdir(parents=True)

    g = bootstrap()

    banner("publishing foo:api and bar:api (same artifactId, different groups)")
    for name in ("foo", "bar"):
        gradle(publisher(name), g, "-q", "publishToMavenLocal",
               f"-Dmaven.repo.local={WORK}/m2").check_returncode()

    banner("running consumer's installDist (expect collision)")
    r = gradle(consumer(), g, "installDist")
    print(r.stdout, r.stderr, sep="")
    if "is a duplicate but no duplicate handling" in r.stdout + r.stderr:
        print("\n\033[1;32m✓ reproduced gradle/gradle#20294\033[0m")
        return 0
    print("\n\033[1;31m✗ did NOT reproduce — Gradle behaviour may have changed\033[0m")
    return 1


if __name__ == "__main__":
    sys.exit(main())
```

Closing as "not planned" four years ago is fine; the issue still
exists. The cost still scales linearly with the registry. 
A metadata flag could at least help matters as a cheap fix.
